vortex-layout/src/layouts/struct_/reader.rs

```diff
@@ -1,27 +1,91 @@
+use std::sync::{Arc, OnceLock};
+
+use vortex_array::aliases::hash_map::HashMap;
 use vortex_array::ContextRef;
-use vortex_error::{vortex_panic, VortexResult};
+use vortex_dtype::{DType, Field, FieldName, StructDType};
+use vortex_error::{vortex_err, vortex_panic, VortexExpect, VortexResult};
 
 use crate::layouts::struct_::StructLayout;
-use crate::{LayoutData, LayoutEncoding, LayoutReader};
+use crate::segments::AsyncSegmentReader;
+use crate::{LayoutData, LayoutEncoding, LayoutReader, LayoutReaderExt};
 
-#[derive(Debug)]
-pub struct StructScan {
+#[derive(Clone)]
+pub struct StructReader {
     layout: LayoutData,
+    ctx: ContextRef,
+
+    segments: Arc<dyn AsyncSegmentReader>,
+
+    field_readers: Arc<[OnceLock<Arc<dyn LayoutReader>>]>,
+
+    field_lookup: HashMap<FieldName, usize>,
+    dtype_index: Box<[DType]>,
 }
 
-impl StructScan {
-    pub(super) fn try_new(layout: LayoutData, _ctx: ContextRef) -> VortexResult<Self> {
+impl StructReader {
+    pub(super) fn try_new(
+        layout: LayoutData,
+        segments: Arc<dyn AsyncSegmentReader>,
+        ctx: ContextRef,
+    ) -> VortexResult<Self> {
         if layout.encoding().id() != StructLayout.id() {
             vortex_panic!("Mismatched layout ID")
         }
 
+        let dtype = layout.dtype();
+        let DType::Struct(struct_dt, _) = dtype else {
+            vortex_panic!("Mismatched dtype {} for struct layout", dtype);
+        };
+
+        let field_readers = struct_dt.names().iter().map(|_| OnceLock::new()).collect();
+
+        let field_lookup = struct_dt
+            .names()
+            .iter()
+            .enumerate()
+            .map(|(i, name)| (name.clone(), i))
+            .collect();
+
+        // TODO(joe): Think about only expanding the fields when we need them -- lazily in the
```

Comment:
Yeah, we can't do this. It forces eager evaluation of all fields, which goes against the point of lazy dtypes